src/compiler/program.ts

```diff
@@ -626,13 +627,29 @@ namespace ts {
                 onlyRecordFailures = !directoryProbablyExists(directory, state.host);
             }
         }
-        return forEach(extensions, tryLoad);
 
-        function tryLoad(ext: string): string {
-            if (ext === ".tsx" && state.skipTsx) {
-                return undefined;
+        if (state.skipTsx)
+            extensions = filter(extensions, ext => ext !== "tsx");
+
+        // First try to keep/add an extension: importing "./foo.ts" can be matched by a file "./foo.ts", and "./foo" by "./foo.d.ts"
+        const keepOrAddExtension = forEach(extensions, ext =>
+           tryLoad(fileExtensionIs(candidate, ext) ? candidate : candidate + ext));
+        if (keepOrAddExtension) {
+            return keepOrAddExtension;
+        }
+
+        // Then try stripping a ".js" or ".jsx" extension and replacing it with a different one, e.g. "./foo.js" can be matched by "./foo.ts" or "./foo.d.ts"

```

Comment:
A recursive definition could mean matching 'foo.js.js.js' with 'foo.ts'. Also, if we remove a '.js' extension from a file, we don't want to try adding it back, or adding '.jsx' in its place. There may be other ways to improve the current code.